app_spec/zomes/blog/code/src/blog.rs

```diff
@@ -319,20 +323,83 @@ pub fn handle_memo_address(content: String) -> ZomeApiResult<Address> {
     hdk::entry_address(&memo_entry(content))
 }
 
-pub fn handle_create_post(content: String, in_reply_to: Option<Address>) -> ZomeApiResult<Address> {
+pub fn create_topics(topics: &Vec<String>) -> ZomeApiResult<Vec<Address>> {
+    let mut topic_addresses = vec![];
+    for topic in topics{
+        let entry = Entry::App("topic".into(), Topic{topic: topic.to_string()}.into());
+        let topic_entry_address = hdk::entry_address(&entry)?;
+        match hdk::get_entry(&topic_entry_address)? {
+            Some(_entry) => {
+                topic_addresses.push(topic_entry_address);
+            },
+            None => {
+                let address = hdk::commit_entry(&entry)?;
+                topic_addresses.push(address);
+            }
+        }
+    };
+    Ok(topic_addresses)
+}
+
+pub fn create_topic_indexs(post: &Address, mut topics: Vec<String>) -> ZomeApiResult<String> {
+    let mut link_combinations = vec![]; //Vector for link combinations on post
+    topics = topics.iter().map(|topic| topic.to_lowercase()).collect(); //put topics to lowercase
+    topics.sort_by(|a, b| b.cmp(&a)); //sort topics to be in decending alphabetical order - allows less links to be made
+
+    for (i, _) in topics.iter().enumerate(){ //Iterate over topics and get all permutations of topics
+        let combinations = topics.iter().combinations(i);
+        for c in combinations.into_iter(){
+            link_combinations.push(c);
+        };
+    };
+    link_combinations.push(topics.iter().collect());
+    link_combinations = link_combinations[1..link_combinations.len()].to_vec();
+    hdk::debug(format!("Topic combination links to be made: {:?}", link_combinations))?;
+    for link in link_combinations{ //Create link combinations for expression indexing
+        let link_strings: Vec<String> = link.iter().map(|link_value| format!("{}", link_value,) ).collect(); //Create link string for each set of topic combinations
+        let link_string = link_strings.join(":");
+        let start_address = hdk::entry_address(&Entry::App("topic".into(), Topic{topic: link[0].to_string()}.into()))?;
+        hdk::link_entries(&start_address, post, "topic_index".to_string(), Some(link_string))?; //create link with type: topic_index and generate link_string
+    };
+    Ok("Link indexs created".to_string())
+}
+
+pub fn handle_create_post(content: String, in_reply_to: Option<Address>, topics: Vec<String>) -> ZomeApiResult<Address> {
     let address = hdk::commit_entry(&post_entry(content))?;
+    create_topics(&topics)?;
+    create_topic_indexs(&address, topics)?;
 
-    hdk::link_entries(&AGENT_ADDRESS, &address, "authored_posts")?;
+    hdk::link_entries(&AGENT_ADDRESS, &address, "authored_posts", None)?;
```

Comment:
I think I was wrong with what I mentioned before. The plan is for every link to have a tag even if it is an empty string. Therefore the last parameter to `hdk::link_entries` cannot be none. 

What I mean to say was that the last parameter for `hdk::get_links` will be optional where the `None` type means to retrieve all links regardless of the tag. 

Sorry about that

Reply:
What about if we still allow passing an option to link_entries but then map None to an empty string. That way tag will be an option the same as it is when using get_links. I think using either option or just a string for both get_links and link_entries functions might make the most sense - but maybe I am missing something that this approach would cause in the deeper implementation @ core.